design/custom-request-timeout.md

```diff
@@ -0,0 +1,256 @@
+- A high-level coverage on the struct changes required in the code.
+- Arguments passed for Envoy API calls.
+
+## Proposed YAML fields
+
+```yaml
+apiVersion: contour.heptio.com/v1beta1
+kind: IngressRoute
+metadata:
+    name: github.com
+    namespace: prod
+spec:
+    virtualhost:
+        fqdn: www.github.com
+        routes:
+        - match: /contour
+          # the timeout applied to the requests in this route
+          timeoutPolicy:
+              # the request timeout applied on this route, it is the maximum time that a client will await
+              # a response to its request.
+              request: 500ms
+              # the maximum time that an idle connection is kept open when there are no active requests
+              idle: 2s
+          retryPolicy:
+              # the number of retries it should make on failure
+              maxRetries: "7"
+              # the conditions for which it should reattempt
+              onStatusCodes:
+              - 501
+              - 502
+              # timeout between the requests, if greater than request timeout, this field is ignored
+              perTryTimeout: 75ms
+        - match: /static-site
+          timeoutPolicy:
+              request: 200ms
+              idle: 1s
+          retryPolicy:
+              maxRetries: 2
+              # for all statusCodes but 504, retry the request
+              onStatusCodes:
+              - 5xx
+              perTryTimeout: 75ms
+```
+
+The naming conventions of the proposed YAML fields are inspired from [HTTP Timeouts](https://tools.ietf.org/id/draft-thomson-hybi-http-timeout-00.html#rfc.section.4).
+
+# Detailed Design
+
+## Reading from YAML
+The ingressroute spec struct will be updated to contain Timeout and Retry struct members
+
+
+```golang
+// contour/apis/contour/v1beta1/ingressroute.go
+
+// Route contains the set of routes for a virtual host
+type Route struct {
+    [... other members ...]
+	// The request timeout for this route
+	Timeout TimeoutPolicy `json:"timeoutPolicy,omitempty"`
+	// The retry attempts for this route
+	Retry RetryPolicy `json:"retryPolicy,omitempty"`
+}
+
+// TimeoutPolicy defines the attributes associated with timeout
+type TimeoutPolicy struct {
+	// Timeout for receiving a response from the server after processing a request from client in milliseconds
+	Request	string `json:"request"`
+	// Timeout for an idle connection to terminate when there are no active requests
+	Idle string `json:"idle"`
+}
+
+// RetryPolicy defines the attributes associated with retrying policy
+type RetryPolicy struct {
+	// MaxRetries is maximum allowed number of retries
+	MaxRetries	int `json:"maxRetries"`
+	// Perform retry on failed requests with the matched status codes or aggregated as 5xx
+	OnStatusCodes []string `json:"onStatusCodes"`
+	// PerTryTimeout specifies the timeout per retry attempt. Ignored if OnStatusCodes are empty
+	PerTryTimeout string `json:"perTryTimeout"`
```

Comment:
If we don't let people provide an empty set of OnStatusCodes then PerTryTimeout will always be consulted. In that case we should call out the default value if not supplied; ie, its reasonable for people to write
````
retryPolicy:
  maxRetries: 6
```
We should make Contour behave sensibly in this instance.

Reply:
Noted. I will mention this more clearly in the follow up for this design doc.